debugger/debugger.js

```diff
@@ -0,0 +1,310 @@
+    this.print("----------------------------------------------------------------------------------\n");
+    this.print(" ==> " + this.get_source_line(lineno - 1) + "\n");
+    this.print("----------------------------------------------------------------------------------\n");
+};
+
+Sk.Debugger.prototype.set_suspension = function(suspension) {
+    var parent = null;
+    if (!hasOwnProperty(suspension, "filename") && suspension.child instanceof Sk.misceval.Suspension) {
+        suspension = suspension.child;
+    }
+        
+    // Pop the last suspension of the stack if there is more than 0
+    if (this.suspension_stack.length > 0) {
+        this.suspension_stack.pop();
+        this.current_suspension -= 1;
+    }
+    
+    // Unroll the stack to get each suspension.
+    while (suspension instanceof Sk.misceval.Suspension) {
+        parent = suspension;
+        this.suspension_stack.push(parent);
+        this.current_suspension += 1;
+        suspension = suspension.child;
+    }
+
+    suspension = parent;
+    
+    this.print_suspension_info(suspension);
+};
+
+Sk.Debugger.prototype.add_breakpoint = function(filename, lineno, colno, temporary) {
+    var key = this.generate_breakpoint_key(filename, lineno, colno);
+    this.dbg_breakpoints[key] = new Sk.Breakpoint(filename, lineno, colno);
+    if (temporary) {
+        this.tmp_breakpoints[key] = true;
+    }
+};
+
+Sk.Debugger.prototype.suspension_handler = function(susp) {
+    return new Promise(function(resolve, reject) {
+        try {
+            resolve(susp.resume());
+        } catch(e) {
+            reject(e);
+        }
+    });
+};
+
+Sk.Debugger.prototype.resume = function() {
+    // Reset the suspension stack to the topmost
+    this.current_suspension = this.suspension_stack.length - 1;
+    
+    if (this.suspension_stack.length === 0) {
+        this.print("No running program");
+    } else {
+        var promise = this.suspension_handler(this.get_active_suspension());
+        promise.then(this.success.bind(this), this.error.bind(this));
+    }
+};
+
+Sk.Debugger.prototype.pop_suspension_stack = function() {
+    this.suspension_stack.pop();
+    this.current_suspension -= 1;
+};
+
+Sk.Debugger.prototype.success = function(r) {
+    if (r instanceof Sk.misceval.Suspension) {
+        this.set_suspension(r);
+    } else {
+        if (this.suspension_stack.length > 0) {
+            // Current suspension needs to be popped of the stack
+            this.pop_suspension_stack();
+            
+            if (this.suspension_stack.length === 0) {
+                this.print("Program execution complete");
+                return;
+            }
+            
+            var parent_suspension = this.get_active_suspension();
+            // The child has completed the execution. So override the child's resume
+            // so we can continue the execution.
+            parent_suspension.child.resume = function() {};
```

Comment:
@rixner 
So suspension stacks are recursive by nature in the way they are being returned from the child. This example below will kind of explain what happens. 

def foo(a, b):
    a = a + 1
    print("foo")

def goo(c, d):
    foo(c+1, d+1)
    print("goo")

def boo(e, f):
    goo(e+1, f+1)
    print("boo")

boo(1, 2)

Now lets say we put a breakpoint in foo(1, 2)... So at this point we have a suspension stack starting from main -> boo -> goo -> foo..

Now once foo completes execution, then we come to goo. At this point, unless we override the resume function for foo, when goo's execution resumes you will see that it again starts executing foo. That is because whenever a suspension resumes it first run's the child's suspension (or resume method). That is what is happening. It's just like a stack call when I am unwinding the stack but apart from that I sort of mark that the child has completed execution so no need to re-execute it. 

I know its a slightly awkward explanation, but let me know if this makes sense..